File: Cmps101/pa1/Lex.java

//Thomas Brochard
//Programming assignment 1
import java.io.*;
import java.util.Scanner;

public class Lex {
  public static void main(String[] args) {
    // Check for arguments
    if (args.length != 2) {
      System.out.println("Usage: Lex [] []");
      System.exit(1);
    }
    // Read input, get String[] and length
    Scanner sc = null;
    PrintWriter out = null;
    try {
      sc = new Scanner(new File(args[0]));
      out = new PrintWriter(new FileWriter(args[1]));
    } catch (IOException e) {
      throw new RuntimeException("File not found: " + args[0] + ", " + args[1]);
    }
    sc.useDelimiter("\\Z");
    String content = sc.next();
    String[] words = content.split("\n");

    // initlialize indeces
    int[] indices = new int[words.length];
    for (int i = 0; i < indices.length; i++)
      indices[i] = i;

    List myList = new List();
    myList.append(0);
    for (int i = 1; i < words.length; i++) {
      myList.moveFront();
      while (myList.index() != -1 && words[i].compareTo(words[myList.get()]) > 0) {
        myList.moveNext();
      }
      if (myList.index() != -1)
        myList.insertBefore(i);
      else
        myList.append(i);
    }

    myList.moveFront();
    while (myList.index() != -1) {
      out.println(words[myList.get()]);
      myList.moveNext();
    }
    sc.close();
    out.close();
  }
}

File: Cmps101/pa2/Lex.c
//-----------------------------------------------------------------------------
// Thomas Brochard
// pa2
// Lex.c
//----------------------------------------------------------------------------
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 250

int main(int argc, char * argv[]) {

  FILE *in, *out;
  char line[MAX_LEN];
  char* words[MAX_LEN];

  // check command line for correct number of arguments
  if (argc != 3) {
    printf("Usage: %s <input file> <output file>\n", argv[0]);
    exit(1);
  }

  // open files for reading and writing
  in = fopen(argv[1], "r");
  out = fopen(argv[2], "w");
  if (in == NULL) {
    printf("Unable to open file %s for reading\n", argv[1]);
    exit(1);
  }
  if (out == NULL) {
    printf("Unable to open file %s for writing\n", argv[2]);
    exit(1);
  }

  int count = 0;
  /* read each line of input file, then count and print tokens*/
  while (fgets(line, MAX_LEN, in) != NULL) {
    words[count] = (char*) malloc(255);
    strcpy(words[count], line);
    count = count + 1;
  }
  int len = count;

  List myList = newList();
  append(myList, 0);
  int i;
  for (i = 1; i < len; i++) {
    moveFront(myList);
    while (index(myList) != (-1) && strcmp(words[i], words[get(myList)]) > 0) {
      moveNext(myList);
    }
    if (index(myList) != (-1)) {
      insertBefore(myList, i);
    } else {
      append(myList, i);
    }
  }

  moveFront(myList);
  while (index(myList) != (-1)) {
    fprintf(out, "%s", words[get(myList)]);
    moveNext(myList);
  }

  freeList(&myList);

  i = 0;
  for (i = 0; i < len; i++) {
    free(words[i]);
  }

  //close files
  fclose(in);
  fclose(out);
  return (0);
}
